zkevm-circuits/src/evm_circuit/execution/memory_copy.rs

```diff
@@ -0,0 +1,532 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::MAX_MEMORY_SIZE_IN_BYTES,
+        step::ExecutionState,
+        table::TxContextFieldTag,
+        util::{
+            constraint_builder::{
+                ConstraintBuilder, StepStateTransition, Transition::Delta,
+            },
+            math_gadget::ComparisonGadget,
+            memory_gadget::BufferGetDataGadget,
+            Cell,
+        },
+        witness::{Block, Call, ExecStep, OpcodeExtraData, Transaction},
+    },
+    util::Expr,
+};
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+
+const MAX_COPY_BYTES: usize = 39;
+
+/// Multi-step gadget for copying data from memory or Tx calldata to memory
+#[derive(Clone, Debug)]
+pub(crate) struct CopyToMemoryGadget<F> {
+    // The src memory address to copy from
+    src_addr: Cell<F>,
+    // The dst memory address to copy to
+    dst_addr: Cell<F>,
+    // The number of bytes left to copy
+    bytes_left: Cell<F>,
+    // The src address bound of the buffer
+    src_addr_end: Cell<F>,
+    // Indicate whether src is from Tx Calldata
+    from_tx: Cell<F>,
+    // Transaction ID, optional, only used when src_is_tx == 1
+    tx_id: Cell<F>,
+    // Buffer get data gadget
+    buffer_get_data:
+        BufferGetDataGadget<F, MAX_COPY_BYTES, MAX_MEMORY_SIZE_IN_BYTES>,
+    // The comparison gadget between num bytes copied and bytes_left
+    finish_gadget: ComparisonGadget<F, 4>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for CopyToMemoryGadget<F> {
+    const NAME: &'static str = "COPYTOMEMORY";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::CopyToMemory;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let src_addr = cb.query_cell();
+        let dst_addr = cb.query_cell();
+        let bytes_left = cb.query_cell();
+        let src_addr_end = cb.query_cell();
+        let from_tx = cb.query_bool();
+        let tx_id = cb.query_cell();
+        let buffer_get_data =
+            BufferGetDataGadget::construct(cb, &src_addr, &src_addr_end);
+
+        let from_memory = 1.expr() - from_tx.expr();
+        let rw_counter = cb.curr.state.rw_counter.expr();
+        let mut rw_counter_offset = 0.expr();
+        // Copy bytes from src and dst
+        for i in 0..MAX_COPY_BYTES {
+            let read_flag = buffer_get_data.read_from_buffer(i);
+            // Read bytes[i] from memory
+            cb.condition(from_memory.clone() * read_flag.clone(), |cb| {
+                cb.memory_lookup_with_counter(
+                    rw_counter.clone() + rw_counter_offset.clone(),
+                    0.expr(),
+                    src_addr.expr() + i.expr(),
+                    buffer_get_data.byte(i),
+                )
+            });
+            rw_counter_offset =
+                rw_counter_offset + from_memory.clone() * read_flag.clone();
```

Comment:
should we add a wrapper as `cb.memory_lookup`? that's also how we do in `cb.rw_lookup`

Reply:
We cannot use `cb.memory_lookup` here because it will increase the `rw_counter` no matter this lookup happens or not, and it will cause gap in the `rw_counter`